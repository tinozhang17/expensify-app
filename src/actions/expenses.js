import moment from 'moment';
import database from '../firebase/firebase';

/* Action Generators */
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = '',
            createdAt = moment().valueOf()
        } = expenseData;
        const expense = { description, note, amount, createdAt };
        // below we are returning another promise, so that it can be further chained later, if needed.
        return database.ref('expenses')
            .push(expense)
            .then((snapshot) => {
             dispatch(addExpense({
                 id: snapshot.key,
                 ...expense
             }));
            });
    };
};

export const removeExpense = (id) => ({
    type: 'REMOVE_EXPENSE',
    expenseId: id
});

export const editExpense = (id, updateObj) => ({
    type: 'EDIT_EXPENSE',
    expenseId: id,
    updateObj
});

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch) => {
        const expenses = [];
        return database.ref('expenses').once('value').then((snapshot) => {
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            dispatch(setExpenses(expenses));
        });
    };
};