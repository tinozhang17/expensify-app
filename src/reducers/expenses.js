// combineReducers allows us to define multiple functions that define how our redux application changes

// Expenses Reducer
const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'SET_EXPENSES':
            return [...state, ...action.expenses];
        case 'ADD_EXPENSE':
            return [...state, action.expense];
        case 'REMOVE_EXPENSE':
            return state.filter((expenseObj) => expenseObj.id !== action.expenseId);
        case 'EDIT_EXPENSE':
            return state.map((expenseObj) => {
                if (expenseObj.id === action.expenseId) {
                    return {
                        ...expenseObj,
                        ...action.updateObj
                    };
                } else {
                    return expenseObj;
                }
            });
        default:
            return state;
    }
};
