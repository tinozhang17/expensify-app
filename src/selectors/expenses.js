import moment from 'moment';

export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const startDateMatch = startDate ? startDate.isSameOrBefore(moment(expense.createdAt), 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(moment(expense.createdAt)) : true;
        const textMatch = text === "" || expense.description.toLowerCase().includes(text.toLowerCase());
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1; // most recent first
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1; // largest amount to smallest amount
        }
    });
};
