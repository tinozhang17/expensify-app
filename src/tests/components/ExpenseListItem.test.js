import React from 'react';
import { shallow } from 'enzyme';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';

test('should render expense list item', () => {
    expect(shallow(<ExpenseListItem {...expenses[1]} />)).toMatchSnapshot();
});