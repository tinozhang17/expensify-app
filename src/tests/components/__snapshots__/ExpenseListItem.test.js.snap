// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list item 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/2"
  >
    <h3>
      heat bill
    </h3>
  </Link>
  <p>
    $5.00
     - 
    December 27th, 1969
  </p>
</div>
`;
